### DISCUSSION TITLE
How to resolve special NuGet version dependency scenario?

### DISCUSSION THREAD

**Discussion by bryan-wu** [adamralph/minver#519] [Q&A] [answered]
# Context

I have the following VS2019 solution structure where Project B has a ProjectReference to ProjectA:
- Solution
       |
       L - - - ProjectA
       |
       L - - - ProjectB

## ProjectA.csproj
```xml
<Project Sdk="My.Sdk">

  <PropertyGroup>
    <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="SomeNuget" Version="1.0.8" />
  </ItemGroup>

</Project>
```

## ProjectB.csproj
```xml
<Project Sdk="My.Sdk">

  <PropertyGroup>
    <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
    <MinVerTagPrefix>special-</MinVerTagPrefix>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\SolutionRoot\ProjectA.csproj" />
  </ItemGroup>

</Project>
```

## Output of `git tag --list`
```
1.0.1
1.0.2
special-1.0.0-preview
```

## ProjectA.nuspec (generated as part of resulting .nupkg via `dotnet pack`)
```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>ProjectA</id>
    <version>1.0.3-alpha.0.36</version>
    <authors>Microsoft</authors>
    <dependencies>
      <group targetFramework=".NETFramework4.6.1">
        <dependency id="SomeNuget" version="1.0.8" exclude="Build,Analyzers" />
      </group>
      <group targetFramework=".NETStandard2.0">
        <dependency id="SomeNuget" version="1.0.8" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
</package>
```

## ProjectB.nuspec (generated as part of resulting .nupkg via `dotnet pack`)
```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>ProjectB</id>
    <version>1.0.0-preview</version>
    <authors>Me</authors>
    <dependencies>
      <group targetFramework=".NETFramework4.6.1">
        <dependency id="ProjectA" version="1.0.3-alpha.0.36" exclude="Build,Analyzers" />
      </group>
      <group targetFramework=".NETStandard2.0">
        <dependency id="ProjectA" version="1.0.3-alpha.0.36" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
</package>
```

# Goal
Create ProjectB NuGet package with .nuspec dependency on the latest valid version of Project A (1.0.2)

## Example of ProjectB.nuspec I want
```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>ProjectB</id>
    <version>1.0.0-preview</version>
    <authors>Me</authors>
    <dependencies>
      <group targetFramework=".NETFramework4.6.1">
        <dependency id="ProjectA" version="1.0.2" exclude="Build,Analyzers" />
      </group>
      <group targetFramework=".NETStandard2.0">
        <dependency id="ProjectA" version="1.0.2" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
</package>
```

# Issue
I'm running the following commands at the solution root directory level:
1. `dotnet restore`
2. `dotnet build --configuration Release --no-restore -p:Deterministic=true`
3. `dotnet pack --configuration Release --no-build`

MinVer increments ProjectA's version to `1.0.3-alpha.0.36` within expected behavior. However, I want ProjectB.nuspec to have a version dependency on the latest version of Project A (1.0.2) instead of `1.0.3-alpha.0.36`.

## What I've tried
1. Using PackageReference instead of ProjectReference
    - This doesn't work since Project A is in a private NuGet feed that Github CI doesn't have access to. 
2. Disabling MinVer for ProjectA via `MinVerSkip`
    - This results in ProjectA having version 1.0.0. However, 1.0.0 is not the latest version (1.0.2) and I shouldn't commit this change to the main Github branch.
3. Setting `<MinVerVersionOverride>1.0.2</MinVerVersionOverride>` in ProjectA.csproj
    - This satisfies the goal, but I haven't been able to find a Condition attribute for the `MinVerVersionOverride` element that would set `MinVerVersionOverride` if the latest git tag contains the "special-" `MinVerTagPrefix`. Perhaps there's another Condition I can check for if I only want the `MinVerVersionOverride` to apply when releasing a new version of ProjectB.
 
## Questions
1. Is there an existing solution to achieve the goal? If so, please connect me to the solution. If not, then would using the MinVerVersionOverride element and a Condition attribute mentioned above be the best practice?
2. Should the MinVer command line tool be used to satisfy the goal?

Thank you for your time.

**Comment by adamralph** [comment, marked as answer]
@bryan-wu this sounds like exactly the same scenario that was described in https://github.com/adamralph/minver/issues/500.

This is the best solution I can think of (from https://github.com/adamralph/minver/issues/500#issuecomment-860248531):

> Let's say you have two projects, `Main` and `Extension`, and `Extension 1.0.0` depends on `Main 1.0.0`. You then start working on `Extension 1.1.0`, and commit changes to that project. As I understand it, the problem you are facing is that, even though you are not committing to `Main`, that project starts to produce versions such as `Main 1.0.1-alpha.0.123` because of the changes your committing to `Extension`, even though you are not touching the `Main` project. You can solve this by "locking" `Main` to `1.0.0` by adding `<MinVerVersionOverride>1.0.0</MinVerVersionOverride>` to `Main.csproj`. Later, when you start committing changes to `Main`, you can "unlock" the version of `Main` by removing that property from `Main.csproj`.

This is a variation of what you mentioned:

> 3. Setting `<MinVerVersionOverride>1.0.2</MinVerVersionOverride>` in ProjectA.csproj
> 
>     * This satisfies the goal, but I haven't been able to find a Condition attribute for the `MinVerVersionOverride` element that would set `MinVerVersionOverride` if the latest git tag contains the "special-" `MinVerTagPrefix`. Perhaps there's another Condition I can check for if I only want the `MinVerVersionOverride` to apply when releasing a new version of ProjectB.

The difference is that _you_ are best placed to know when changes being committed to the repo affect project A and when they don't. As long as they don't, you leave the `MinVerVersionOverride` in place. As soon as they do, you remove it and allow MinVer to start incrementing the version again.
